File: backend/app/schemas/notifications.py
"""
Notification System Pydantic Schemas
"""

from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List, Dict, Any
from datetime import datetime, time


class NotificationPreferencesBase(BaseModel):
    email_enabled: bool = True
    sms_enabled: bool = False
    push_enabled: bool = True
    in_app_enabled: bool = True
    booking_confirmations: bool = True
    booking_reminders: bool = True
    payment_receipts: bool = True
    promotional_offers: bool = True
    loyalty_updates: bool = True
    service_updates: bool = True
    email_address: Optional[str] = None
    phone_number: Optional[str] = None
    preferred_language: str = 'en'
    quiet_hours_start: Optional[time] = None
    quiet_hours_end: Optional[time] = None
    timezone: str = 'UTC'


class NotificationPreferencesResponse(NotificationPreferencesBase):
    id: str
    user_id: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class NotificationBase(BaseModel):
    notification_type: str = Field(..., pattern="^(email|sms|push|in_app)$")
    title: str = Field(..., max_length=500)
    message: str
    event_type: Optional[str] = None
    reference_type: Optional[str] = None
    reference_id: Optional[str] = None
    data: Optional[Dict[str, Any]] = None
    priority: str = Field(default="normal", pattern="^(low|normal|high|urgent)$")


class NotificationCreate(NotificationBase):
    user_id: str


class NotificationResponse(NotificationBase):
    id: str
    user_id: str
    status: str
    sent_at: Optional[datetime] = None
    delivered_at: Optional[datetime] = None
    read_at: Optional[datetime] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class NotificationUpdate(BaseModel):
    read_at: Optional[datetime] = None


class NotificationStats(BaseModel):
    total_notifications: int
    unread_count: int
    by_type: Dict[str, int]
    recent_notifications: List[NotificationResponse]

File: src/lib/api/notifications.ts
/**
 * Notification System API Client
 */

import api from './axios';

export interface Notification {
  id: string;
  user_id: string;
  notification_type: 'email' | 'sms' | 'push' | 'in_app';
  title: string;
  message: string;
  event_type?: string;
  reference_type?: string;
  reference_id?: string;
  data?: any;
  status: string;
  priority: 'low' | 'normal' | 'high' | 'urgent';
  sent_at?: string;
  delivered_at?: string;
  read_at?: string;
  created_at: string;
}

export interface NotificationPreferences {
  id: string;
  user_id: string;
  email_enabled: boolean;
  sms_enabled: boolean;
  push_enabled: boolean;
  in_app_enabled: boolean;
  booking_confirmations: boolean;
  booking_reminders: boolean;
  payment_receipts: boolean;
  promotional_offers: boolean;
  loyalty_updates: boolean;
  service_updates: boolean;
  email_address?: string;
  phone_number?: string;
  preferred_language: string;
  timezone: string;
  created_at: string;
  updated_at: string;
}

export interface NotificationStats {
  total_notifications: number;
  unread_count: number;
  by_type: Record<string, number>;
  recent_notifications: Notification[];
}

class NotificationService {
  async getPreferences(): Promise<NotificationPreferences> {
    const response = await api.get<NotificationPreferences>('/notifications/preferences');
    return response.data;
  }

  async updatePreferences(data: Partial<NotificationPreferences>): Promise<NotificationPreferences> {
    const response = await api.put<NotificationPreferences>('/notifications/preferences', data);
    return response.data;
  }

  async getNotifications(params?: {
    unread_only?: boolean;
    limit?: number;
    offset?: number;
  }): Promise<Notification[]> {
    const queryParams = new URLSearchParams();
    if (params?.unread_only !== undefined) queryParams.append('unread_only', params.unread_only.toString());
    if (params?.limit) queryParams.append('limit', params.limit.toString());
    if (params?.offset) queryParams.append('offset', params.offset.toString());

    const response = await api.get<Notification[]>(`/notifications?${queryParams.toString()}`);
    return response.data;
  }

  async markAsRead(notificationId: string): Promise<Notification> {
    const response = await api.patch<Notification>(`/notifications/${notificationId}/read`, {});
    return response.data;
  }

  async markAllAsRead(): Promise<void> {
    await api.post('/notifications/mark-all-read', {});
  }

  async getStats(): Promise<NotificationStats> {
    const response = await api.get<NotificationStats>('/notifications/stats');
    return response.data;
  }
}

export const notificationService = new NotificationService();
